### ISSUE TITLE
PrismaClient with `--data-proxy` throws ambiguous error on transaction timeout

### ISSUE THREAD

**Issue by rtbenfield** [prisma/prisma#18710]
### Bug description

When using the PrismaClient with `--data-proxy`, transaction timeouts will result in an unexpected error message when the following conditions are met.
1. The runtime environment is detected as "browser", utilizing the built-in `fetch`
2. The transaction timeout expires between the last Prisma operation and the commit / rollback step

### How to reproduce

I've written a minimal reproduction [here](https://github.com/rtbenfield/repro-prisma-dp-transaction) with test cases.

With the above conditions met, running the following will throw an error with the message "This request could not be understood by the server".
```ts
// BAD
await prisma.$transaction(
  async (tx) => {
    await new Promise((resolve) => setTimeout(resolve, 1_100));
  },
  { timeout: 1_000 }
);
```

However, if a Prisma query or mutation is introduced after the timeout expires, but before the commit, the expected errors is thrown.
```ts
// GOOD
await prisma.$transaction(
  async (tx) => {
    await new Promise((resolve) => setTimeout(resolve, 1_100));
    await tx.someData.count(); // <- throws here with "Transaction already closed"
  },
  { timeout: 1_000 }
);
```

### Expected behavior

The transaction should throw, but with the message:
> Transaction API error: Transaction already closed

### Prisma information

```prisma
generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider  = "postgresql"
  url       = env("DATA_PROXY_URL")
  directUrl = env("DATABASE_URL")
}

model SomeData {
  id String @id @default(cuid())
}

```

```ts
import { PrismaClient } from "@prisma/client/edge";

const prisma = new PrismaClient({
  log: ["error", "info", "query", "warn"],
});

await prisma.$transaction(
  async (tx) => {
    await new Promise((resolve) => setTimeout(resolve, 1_100));
  },
  { timeout: 1_000 }
);
```

### Environment & setup

- OS: WSL Ubuntu
- Database: PostgreSQL
- Node.js version: v19.8.1

### Prisma Version

```
prisma                  : 4.12.0
@prisma/client          : 4.12.0
Current platform        : debian-openssl-3.0.x
Query Engine (Node-API) : libquery-engine 659ef412370fa3b41cd7bf6e94587c1dfb7f67e7 (at node_modules/@prisma/engines/libquery_engine-debian-openssl-3.0.x.so.node)
Migration Engine        : migration-engine-cli 659ef412370fa3b41cd7bf6e94587c1dfb7f67e7 (at node_modules/@prisma/engines/migration-engine-debian-openssl-3.0.x)
Format Wasm             : @prisma/prisma-fmt-wasm 4.12.0-67.659ef412370fa3b41cd7bf6e94587c1dfb7f67e7
Default Engines Hash    : 659ef412370fa3b41cd7bf6e94587c1dfb7f67e7
Studio                  : 0.483.0
```

**Comment by rtbenfield** [comment]
With this isolated to only "browser" environment using built-in `fetch`, I think the bug is that the response stream is being read twice in the commit/rollback transaction flow.

For most requests, `requestInternal` will follow [this flow](https://github.com/prisma/prisma/blob/main/packages/engine-core/src/data-proxy/DataProxyEngine.ts#L347:L365):
request ▶️ responseToError ▶️ handleError ▶️ json

But transaction commit/rollback follows [this flow](https://github.com/prisma/prisma/blob/main/packages/engine-core/src/data-proxy/DataProxyEngine.ts#L443:L456):
request ▶️ json ▶️ responseToError ▶️ handleError

`responseToError` will (for non-2xx status codes) call `getResponseErrorBody`, which reads the response as text [here](https://github.com/prisma/prisma/blob/304c54c732921c88bfb57f5730c7f81405ca83ea/packages/engine-core/src/data-proxy/errors/utils/responseToError.ts#L51). Once that stream is consumed, it will fail on any additional attempts to read the body again.

For most workflows, `responseToError` will throw before reaching the `json()` call. The stream is either consumed as text by `getResponseErrorBody` or as JSON by `requestInternal`, but not both. 

The transaction commit/rollback is consuming the JSON unconditionally first, so it can't be read again as text in `getResponseErrorBody`.

TLDR; flipping the order of operations [here](https://github.com/prisma/prisma/blob/304c54c732921c88bfb57f5730c7f81405ca83ea/packages/engine-core/src/data-proxy/DataProxyEngine.ts#L449:L456) will probably do the trick 🙂

**Comment by FGoessler** [comment]
Hi,

We’ve noticed there hasn’t been activity on this issue for quite some time, and we suspect it might no longer be relevant. As part of our ongoing effort to streamline open issues in this repository, we’re closing this one to focus on what’s most important to our community.

If you believe this issue is still relevant and requires attention, feel free to reopen it. Reopening the issue is a valuable signal to us about what matters most to you.

You can learn more about our approach and motivations in the [Prisma ORM Manifesto](https://www.prisma.io/blog/prisma-orm-manifesto#1-defining-first-class-databases).

With love,
Your Prisma ORM Team ❤️
